source/agora/consensus/validation/Transaction.d

```diff
@@ -76,6 +79,15 @@ public string isInvalidReason (
         if (!sum_unspent.add(utxo_value.output.value))
             return "Transaction: Input overflow";
 
+        // note: this is strictly not necessary to be here, the Input's script
+        // should be evaluated for validity and then the Input could be kept
+        // until the unlock height becomes valid. Alternatively we could reject
+        // it here right away and force the party to send it again at the right
+        // time. However, we run into a risk if we ever implement caching of
+        // rejecting them if they're submitted too early.
+        if (height < utxo_value.unlock_height + input.unlock_age)
```

Comment:
I think there's not enough testing on this, but it all looks good.

Reply:
Can you explain? The unittests are not enough?

Reply:
This is a question,
are you thinking of replacing the `utxo_value.unlock_height` with the `height_lock` later?
In my Think Verification methods are confusing when it is a Freezing transaction or a melting transaction.

Reply:
This code you're commenting on does not use `height_lock`. This is the transaction-level lock. 

The code instead `unlock_age`, which is an input-level lock. Unlock age is the number of externalized blocks between when a UTXO is first ready to be spent and the current block height.

> In my Think Verification methods are confusing when it is a Freezing transaction or a melting transaction.

Yes it's a little bit complicated right now because we have this special freeze transaction.

But we cannot build the flash layer with just freeze transactions. We need transaction-level and input-level time locks too. This is issue #243.

Reply:
#243 will be part of my demo, so this will all be explained there.